src/sdk/pynni/nni/compression/torch/builtin_quantizers.py

```diff
@@ -29,23 +42,162 @@ class QAT_Quantizer(Quantizer):
+            nudged_zero_point = qmin
+        elif initial_zero_point > qmax:
+            nudged_zero_point = qmax
+        else:
+            nudged_zero_point = torch.round(initial_zero_point)
+
+        op.scale = scale
+        op.zero_point = nudged_zero_point
+
+    def quantize(self, bits, op, real_val):
+        """
+        quantize real value.
+
+        Parameters
+        ----------
+        bits : int
+            quantization bits length
+        op : torch.nn.module
+            target module
+        real_val : float
+            real value to be quantized
+        Returns
+        -------
+        quantized_val : float
+        """
+        transformed_val = op.zero_point + real_val / op.scale
+        qmin = 0
+        qmax = 1 << bits - 1
+        clamped_val = torch.clamp(transformed_val, qmin, qmax)
+        quantized_val = torch.round(clamped_val)
+        return quantized_val
+
+    def dequantize(self, op, quantized_val):
+        """
+        dequantize quantized value.
+
+        Parameters
+        ----------
+        op : torch.nn.module
+            target module
+        quantized_val : float
+            quantized_val value to be dequantized
+
+        Returns
+        -------
+        real_val : float
+        """
+        real_val = op.scale * (quantized_val - op.zero_point)
+        return real_val
+
+    def quantize_weight(self, weight, config, op, **kwargs):
+        if config['weight_bits'] <= 1:
             return weight
-        a = torch.min(weight)
-        b = torch.max(weight)
-        n = pow(2, config['q_bits'])
-        scale = (b-a)/(n-1)
-        zero_point = a
-        out = torch.round((weight - zero_point)/scale)
-        out = out*scale + zero_point
-        orig_type = weight.dtype
-        return out.type(orig_type)
+        # if config['quant_delay'] > self._steps:
+        #     return weight
+        rmin, rmax = self.fixed_range_check(weight)
+        self.update_quantization_param(config['weight_bits'], op, rmin, rmax)
+        out = self.quantize(config['weight_bits'], op, weight)
+        out = self.dequantize(op, out)
+        return out
+
+    def quantize_activation(self, activation, config, op, **kwargs):
+        if config['activation_bits'] <= 1:
+            return activation
+        # if config['quant_delay'] > self._steps:
+        #     return activation
+        # TODO activation dynamic set a, b
+        rmin, rmax = self.fixed_range_check(activation)
+        self.update_quantization_param(config['activation_bits'], op, rmin, rmax)
+        out = self.quantize(config['activation_bits'], op, activation)
+        out = self.dequantize(op, out)
+        return out
+
+    def fold_bn(self, config, **kwargs):
+        # TODO simulate folded weight
+        pass
+
+    def step(self):
+        """override compressor step method, update _step attribute, quantization only happens after certain number of steps
+        """
+        self.steps += 1
```

Comment:
where do you use `self.steps`?

Reply:
`self.steps` is used to track the current steps of model, for example if `quant_delay` is 4000, then quantization will only happen after 4000 steps.